### PR TITLE
ENG-4686 add basic liquibase capability to multitenant

### PR DESCRIPTION
**Pull request by entgigi**

### CODE REVIEW

File: engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantDataAccessor.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import javax.sql.DataSource;
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TenantDataAccessor {
+
+    private static final Logger logger = LoggerFactory.getLogger(TenantDataAccessor.class);
```
Comment (zonia3000) [inline engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantDataAccessor.java]: Can we use `@Slf4j` here?

File: engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantService.java [hunk 2]
```diff
@@ -64,6 +70,48 @@ public List<TenantDto> getTenants() {
...
+    private Map<String, TenantStatus> fetchByFilteringConfig(Map<String, TenantStatus> statuses, boolean isMandatory) {
+        return statuses.keySet().stream()
+                .flatMap(code -> tenantManager.getConfig(code).stream())
+                .filter(tc -> isMandatory == tc.isInitializationAtStartRequired())
+                .map(TenantConfig::getTenantCode)
+                .collect(Collectors.toMap(k -> k, statuses::get));
+    }
+
+    private TenantStatsDto calculate(Map<String, TenantStatus> statuses) {
+        TenantStatsDto d = new TenantStatsDto(0,0,0,0,0);
```
Comment (zonia3000) [inline engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantService.java]: Any reason here to not use `@NoArgsConstructor` and simply instantiate this using `new TenantStatsDto()`? Looking at this constructor it's not easy to understand which parameter populates which field.

File: portal-ui/src/test/java/org/entando/entando/aps/system/services/controller/control/TestTenantController.java [hunk 3]
```diff
@@ -125,10 +141,23 @@ public RequestContext createExtRequestContext(String serverName, ApplicationCont
     private static void recreateTenantManager(String tenants) throws Exception {
         ApplicationContext applicationContext = BaseTestCase.getApplicationContext();
         DefaultSingletonBeanRegistry registry = (DefaultSingletonBeanRegistry) applicationContext.getAutowireCapableBeanFactory();
-        registry.destroySingleton("tenantManager");
+        //registry.destroySingleton("tenantManager");
```
Comment (zonia3000) [inline portal-ui/src/test/java/org/entando/entando/aps/system/services/controller/control/TestTenantController.java]: cleanup

File: engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java [hunk 4]
```diff
@@ -52,6 +54,12 @@ public void init() throws Exception {
         logger.debug("{} (cache info service initialized) ready", this.getClass().getName());
     }
 
+    @Override
+    public void releaseTenantAware() {
+        // FIXME maybe should optimize this destroy
```
Comment (zonia3000) [inline engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java]: ?

File: cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/contentpagemapper/ContentPageMapperManager.java [hunk 5]
```diff
@@ -42,16 +44,26 @@ public class ContentPageMapperManager extends AbstractService implements IConten
...
-		this.getCacheWrapper().initCache(this.getPageManager(), this.getPageModelManager());
+		initTenantAware();
 		logger.debug("{} ready.", this.getClass().getName());
 	}
     
     @Override
     protected void release() {
-        ((AbstractCacheWrapper) this.getCacheWrapper()).release();
+		releaseTenantAware();
         super.release();
```
Comment (zonia3000) [inline cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/contentpagemapper/ContentPageMapperManager.java]: indentation

File: engine/src/main/java/com/agiletec/aps/system/services/keygenerator/KeyGeneratorManager.java [hunk 6]
```diff
@@ -34,16 +35,26 @@ public class KeyGeneratorManager extends AbstractService implements IKeyGenerato
...
+		initTenantAware();
 		logger.debug("{} ready. : last loaded key {}", this.getClass().getName(), this.getCacheWrapper().getUniqueKeyCurrentValue());
 	}
     
     @Override
     protected void release() {
-        this.getCacheWrapper().release();
-        super.release();
+        releaseTenantAware();
+		super.release();
```
Comment (zonia3000) [inline engine/src/main/java/com/agiletec/aps/system/services/keygenerator/KeyGeneratorManager.java:45]: indentation

File: engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantDataAccessor.java (hunk 1 above)
Comment (entgigi) [reply engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantDataAccessor.java]: done

File: engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantService.java (hunk 2 above)
Comment (entgigi) [reply engine/src/main/java/org/entando/entando/aps/system/services/tenants/TenantService.java]: done

File: engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java (hunk 4 above)
Comment (entgigi) [reply engine/src/main/java/org/entando/entando/aps/system/services/cache/CacheInfoManager.java]: done

File: engine/src/main/java/com/agiletec/aps/system/services/keygenerator/KeyGeneratorManager.java (hunk 6 above)
Comment (entgigi) [reply engine/src/main/java/com/agiletec/aps/system/services/keygenerator/KeyGeneratorManager.java:45]: done

File: cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/contentpagemapper/ContentPageMapperManager.java (hunk 5 above)
Comment (entgigi) [reply cms-plugin/src/main/java/com/agiletec/plugins/jacms/aps/system/services/contentpagemapper/ContentPageMapperManager.java]: done

**Review by zonia3000 — APPROVED** [review]

**Review by eugeniosant — APPROVED** [review]

**Comment by entgigi** [comment]
As discussed with the team to remove the code duplication it's not possible without a heavy refactor, so we agreed with @eugeniosant to however merge the pr

**Pull request merged**